### PR TITLE
User attribute value length extension

### LINKED ISSUE #9758: User attributes with a text more than 255 characters
### Description

we need to have user attributes with a text more than 255 characters.

### Discussion

_No response_

### Motivation

we want to have user attributes with a text more than 255 characters.

### Details

Implementation: 
when creating user attributes, the value field is limited to 255 characters, we want to remove this limitation by changing the type of the field "value" in the table "user_attribute" from varchar(255) to TEXT.

### PR DESCRIPTION
**Pull request by vramik**
Closes #9758

The PR adds a possibility to store and search by long user attribute values. It works for both regular and federated user attributes.

For both tables it uses two new columns in the database schema. One BLOB where the actual value is stored, second with its hash. On the hash column there is also an index created to allow effective search.

This approach could lead to very low chance that there would occur conflicts within the hash value, due to that there has been requirement to verify that user actually has indeed the specified value before returning it.

This check has some drawbacks:

1. there is no way to perform the check on count queries as the result from the database contains the number for records fulfilling the criteria
2. similarly there is no way to perform the check when calling `UserAttributeFederatedStorage#getUsersByUserAttributeStream` as it returns only user IDs (`Stream<String>`)
3. effectively changing the characteristics for `UserEntity` - `UserAttributeEntity` associations from LAZY to EAGER as for each returned user its attributes needs to be loaded as well. 

Ad. 3. for most scenarios we load attributes anyway as they are needed, but in case calling `UsersResource#getUsers` with `briefRepresentation` set to true (by default it is false) in has the mentioned impact

### CODE REVIEW

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:995]: Is it OK to throw JAX-RS exceptions from the storage layer? This one will also leak internal details (ok, just the class name but still :)).

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java
Comment (pedroigor) [inline model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java:83]: Would make sense to check whether both values are set to identify more easily an unexpected state where both values are set? I know it is unlike to happen but perhaps as an additional check. Not necessarily a Java assertion but perhaps an `IllegalStateException`.

**Comment by pedroigor** [comment]
What about `FED_USER_ATTRIBUTE`?

File: model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java:995]: Agreed, I believe the correct way is to throw a `ModelException` or something similar and then the service layer should translate. On the other hand I am not sure what we are currently doing in the service layer, I remember we don't have a consistent approach to exceptions

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java:83]: +1

**Review by pedroigor — DISMISSED** [review]

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java
Comment (sguilhen) [inline model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java:84]: May I just suggest to add the attribute name in the message? Something like "User with id %s should not have set both `value` and `long_value`  for attribute %s", user.getId(), name ?

File: model/jpa/src/main/java/org/keycloak/storage/jpa/entity/FederatedUserAttributeEntity.java
Comment (sguilhen) [inline model/jpa/src/main/java/org/keycloak/storage/jpa/entity/FederatedUserAttributeEntity.java:85]: Same thing here regarding the attribute name - if something goes wrong, the name helps to narrow down the faulty attribute.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-24.0.0.xml
Comment (cgeorgilakis) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-24.0.0.xml:22]: Is it possible this to be different per database?
PostgreSQL has field text which is unlimited and with very good search performance.
We already use text field in database to overcome User attribute limit.

Unfortunately, I do not know if something similar exists in other databases.

Comment (vramik) [reply model/jpa/src/main/resources/META-INF/jpa-changelog-24.0.0.xml:22]: AFAIK liquibase already handles datatypes differently for each database. For PostgreSQL it seems it is indeed a text field: https://github.com/liquibase/liquibase/blob/043acd1581b695ad19c484ee985c6283666040d0/liquibase-standard/src/main/java/liquibase/datatype/core/ClobType.java#L104

Would it suffice for your use case?

Comment (cgeorgilakis) [reply model/jpa/src/main/resources/META-INF/jpa-changelog-24.0.0.xml:22]: Yes. Nice.

File: model/jpa/src/main/java/org/keycloak/storage/jpa/JpaUserFederatedStorageProvider.java
Comment (ahus1) [inline model/jpa/src/main/java/org/keycloak/storage/jpa/JpaUserFederatedStorageProvider.java:169]: Reading the comment, it looks like we previously supported searching for attributes longer than 255 characters in the federated attributes, and the code has changed to no longer support this. This would then be a regression. It is also difficult to explain to users why user attribute search lengths are different in both places.

Nitpick: I didn't find the method referenced here. It seems to me that the check is done in `searchForUserByUserAttributeStream`, therefore the comment might needs updating.

```suggestion
            //this should never happen as the exception should be thrown already by JpaUserProvider#searchForUserByUserAttributeStream, but leaving it here as additional check
```

**Review by ahus1 — DISMISSED** [review]
To prevent a denial of service attempt, we should have a maximum length of attributes defined. 

We discussed that the declarative user profile could have a configurable length of an attribute. For the time being and also when UP is disabled, we need to have a (possibly hard coded) maximum length of user attributes.

**Comment by vramik** [comment]
@ahus1 I've kept the PR separated into two commit where the second one contains the requested collision check. PTAL, I'll squash the commits after the review. Thank you.

**Comment by ahus1** [comment]
I'm still reviewing this. I found that in the existing code in the user attribute search sometimes puts everything first to lowercase before searching. When search just for a single attribute, it doesn't. And comparing two strings after making them lowercase is different from comparing two strings and ignoring the case. Even worse, the results can depend on the locale.

I started to prepare some changes locally and will continue reviewing and preparing. Please don't push additional changes while I'm doing this.

**Comment by vramik** [comment]
Thank you ahus1!

> I'm still reviewing this. I found that in the existing code in the user attribute search sometimes puts everything first to lowercase before searching. When search just for a single attribute, it doesn't.

That's a good catch! Those should be unified, thank you for spotting this.

> And comparing two strings after making them lowercase is different from comparing two strings and ignoring the case. Even worse, the results can depend on the locale.

You're right, I didn't realize there are some cases where the outcome could differ. Just a note there are more places which may or may not have the same issue. It's worth checking. I've found e.g. https://github.com/keycloak/keycloak/blob/13fa851a9f0e079aa16ae01d59f0704fefc553c1/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L901 or https://github.com/keycloak/keycloak/blob/13fa851a9f0e079aa16ae01d59f0704fefc553c1/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L957

> I started to prepare some changes locally and will continue reviewing and preparing. Please don't push additional changes while I'm doing this.

+1

**Comment by ahus1** [comment]
> Those should be unified, thank you for spotting this.

Not sure about that, it might break KC for existing users. I'm planning to add two hashes. A future PR might then unify the behavior.

**Comment by ahus1** [comment]
Suggestion after talking to @stianst: 

* Length validation on all unmanaged attributes to have the max as before (255 chars).
* Length validation on all managed attributes to have the max as before (255 chars) unless they have the Keycloak length validator present int their settings. 

This is to make (a) the default configurations of existing migrated realms secure and (b) avoid insecure mis-configurations

**Comment by sschu** [comment]
@ahus1 Do I understand this right that supporting "case-insensitive" search using hash columns means it will not be possible to do fuzzy searches on attributes stored in the USER_ATTRIBUTE table in contrast to firstname/lastname/username/email which are stored in the USER_ENTITY table?

**Comment by ahus1** [comment]
> Do I understand this right that supporting "case-insensitive" search using hash columns means it will not be possible to do fuzzy searches on attributes stored in the USER_ATTRIBUTE table in contrast to firstname/lastname/username/email which are stored in the USER_ENTITY table?

Yes, this is correct. A fuzzy search in the attributes column was never supported in earlier releases, and is beyond the issue and this PR.

**Comment by sschu** [comment]
I know agree it wasn't supported. However, since the user profile hides the implementation detail of where attributes are stored, people might expect this to work similarly to properties today. That's why I am wondering whether introducing this limitation is a good idea or if there are other options (like taking DB collation into account).

**Comment by ahus1** [comment]
> That's why I am wondering whether introducing this limitation is a good idea or if there are other options (like taking DB collation into account).

We can do only one step at a time. This removes the limitation on the values stored, which is already a delicate task that isn't finished yet. Feel free to raise an enhancement request.

**Comment by ahus1** [comment]
Added two more commit: 

* MariaDB and MySQL where using a `TEXT` column with up to 64k on `FED_USER_ATTRIBUTES`. There is now a migration for those values. 
* There is now a length validation that restricts both managed and unmanaged attributes to 2048 characters. This is both larger than the original value of 255 (user attributes) and 2024 (federated attributes), and should suffice for most setups. 
* There is an additional migration guide for MariaDB and MySQL. Having unlimited length values for those databases is not an option for federated attributes.

**Comment by ahus1** [comment]
@pedroigor - I found that when using LDAP, people were able to use longer user attributes (for example for photos) even before. The maximum length validation for unmanaged LDAP attributes applies here as well, so people who are using attribute values in LDAP which are larger than 2048 characters (ca. 1500 for binary data) will need to make those attributes managed and add a length validation. 

I assume this is ok as no user should post an unlimited amount of data to a field in LDAP.

Tests are now green again, and docs updated.

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @vramik and @ahus1. I added a few comments. Maybe I am nitpicking too much, sorry for that. 

I also created a commit to refactor the lambda for hash collision check: https://github.com/mhajas/keycloak/commit/14d9e2fe5092fe1d34cdb9de9a69a61571dd51c5 but I don't want to just push it as I am not sure it is much better, feel free to add it if you like it.

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java:67]: I hope that Keycloak will eventually provide better case-insensitive search, and this the route of collation in databases. Once we want to get that right, we'll probably drop the `LONG_VALUE_HASH_LOWER_CASE` column and replace it with something different. 

So I'm advocating to keep it as is for now, as those changes are beyond the scope of this PR.

File: model/jpa/src/main/java/org/keycloak/storage/jpa/JpaHashUtils.java
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/storage/jpa/JpaHashUtils.java:58]: A default locale can change during a migration from one system to another, or an upgrade in an unexpected way. With the current setup we can't detect when a locale has changed, and we don't offer a migration path from one locale to another. 

With the setup above, there is at least the guarantee that if you enter the exact same input as before in the same upper/lower case, it will always return that as a match. It might miss out on some variants due to the locale. 

A future version of Keycloak could make this configurable, and could include a re-calculation of the hashes. It would need to iterate over all entries in the table, re-calculate the hashes, and then updating those which differ (which would hopefully be only a subset).

File: docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc
Comment (ahus1) [reply docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:1]: added in next commit.

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/UserAttributeEntity.java:67]: Thanks @ahus1. I agree with that.

File: model/jpa/src/main/java/org/keycloak/storage/jpa/JpaHashUtils.java
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/storage/jpa/JpaHashUtils.java:58]: I don't think it is feasible to recalculate the hashes in these tables but I don't see any other option. I like the idea of dropping the lowercase search in the future which would solve this.

Thanks for the explanation!

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/FederatedUserAttributeTextColumnMigration.java
Comment (ahus1) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/FederatedUserAttributeTextColumnMigration.java:43]: Unicode character have sometimes a strange behavior when it comes to lengths, and you're right that I should check this. See below for an example:

https://stackoverflow.com/questions/64213394/string-encoding-with-emoji-in-java


```java
System.out.println("🇻🇺".length())
// -> 4
```

```
SELECT length("🇻🇺"), char_length("🇻🇺")
FROM dual;
-- -> 8, 2
```

I didn't find an equivalent for Java's `length()` in MySQL, so I'll do the following: 

* Change the SQL to use `LENGTH()` to select possibly more rows than necessary.
* Add additional filtering in Java to ensure I'm updating only those where Java's string length is longer.

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java
Comment (sguilhen) [inline server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java:223]: Can't this whole thing be a simplified a bit by just returning from here instead of breaking out of two loops just to skip the rest of the method? i.e. remove the flag, and return if the id matches here.

**Review by sguilhen — CHANGES_REQUESTED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java
Comment (sguilhen) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java:542]: Also, a question: shouldn't this size of attribute fail the length check? Or is a length validator added somewhere that I'm missing?

Comment (ahus1) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java:542]: AFAIK this creates a lambda that runs directly on the server and accesses the model. This doesn't go via the declarative user profile layer, so it can write any data it wants.

File: server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java
Comment (ahus1) [reply server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java:223]: Yes, that work work. I'll add anther commit. The data is in-memory IMHO already, anyway.

Comment (ahus1) [reply server-spi-private/src/main/java/org/keycloak/userprofile/DefaultAttributes.java]: Commit added. Build will hopefully be green. Ready for a final round of review?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java
Comment (sguilhen) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageTest.java:542]: ah, true.. thanks for the info!

**Review by sguilhen — APPROVED** [review]
Thanks @ahus1 ! I'm assuming the commits will be squashed upon merging - otherwise I would request you to fix the commit message that says "Fimpmlified method with a direct return" :)

Approving as long as CI keeps green

**Review by mhajas — APPROVED** [review]
Thank you @vramik and @ahus1! The changes lgtm! I believe this is ready for merging. I will leave this to you as I am only on mobile phone and maybe I am missing something.

**Comment by ahus1** [comment]
> otherwise I would request you to fix the commit message that says "Fimpmlified method with a direct return" :)

Oh, covfefe!!

**Comment by ahus1** [comment]
Thank you for everyone who helped with this task in the discussions, ideas and iterations until we got to this point to get this merged. I hope we got it correct in all necessary the variant where attributes are used. Let's keep an eye on this in the coming weeks before the next KC24 to spot unexpected regressions.

Parts of this change wouldn't have been possible without the declarative user profile. With that in place, it was possible to have both a default length validation and a UI to configure the length of each field. Thank you!

**Comment by ahus1** [comment]
@overheadhunter - no, it didn't. Feel free to open a new issue for that if none exists.

**Pull request merged**
